monarchmoney/monarchmoney.py

```diff
@@ -635,6 +635,120 @@ async def get_cashflow_summary(
+        """
+        Updates a single existing transaction as identified by the transaction_id
+        Explicitly defining parameters that worked in testing.
+        date in "%Y-%m-%d" or 2023-10-30
+        mm.update_transaction(
+            transaction_id="160820461792094418",
+            category_id="160185840107743863",
+            merchant_id="Amazon",
+            goal_id="160826408575920275",
+            amount=123.45,
+            date="2023-11-09",
+            hide_from_reports=False,
+            needs_review=False,
+            notes=f'Updated On: {datetime.now().strftime("%m/%d/%Y %H:%M:%S")}',
+        )
+        """
+        query = gql(
+            """
+        mutation Web_TransactionDrawerUpdateTransaction($input: UpdateTransactionMutationInput!) {
+            updateTransaction(input: $input) {
+            transaction {
+                id
+                amount
+                pending
+                date
+                hideFromReports
+                needsReview
+                reviewedAt
+                reviewedByUser {
+                id
+                name
+                __typename
+                }
+                plaidName
+                notes
+                isRecurring
+                category {
+                id
+                __typename
+                }
+                goal {
+                id
+                __typename
+                }
+                merchant {
+                id
+                name
+                __typename
+                }
+                __typename
+            }
+            errors {
+                ...PayloadErrorFields
+                __typename
+            }
+            __typename
+            }
+        }
+
+        fragment PayloadErrorFields on PayloadError {
+            fieldErrors {
+            field
+            messages
+            __typename
+            }
+            message
+            code
+            __typename
+        }
+        """
+        )
+
+        variables = {
+            "input": {
+                "id": transaction_id,
+            }
+        }
+
+        # Using 'is not None' to allow writing of empty strings
+        if category_id is not None:
+            variables["input"].update({"category": category_id})
+        if merchant_name is not None:
+            variables["input"].update({"name": merchant_name})
+        if goal_id is not None:
+            variables["input"].update({"goalId": goal_id})
+        if date is not None:
+            variables["input"].update({"date": date})
+        if amount is not None:
+            variables["input"].update({"amount": amount})
+        if hide_from_reports is not None:
+            variables["input"].update({"hideFromReports": hide_from_reports})
+        if needs_review is not None:
+            variables["input"].update({"needsReview": needs_review})
+        if notes is not None:
+            variables["input"].update({"notes": notes})
```

Comment:
With testing I found that you can never pass "amount" or "date" as None or an empty string.  Therefore we only append to variables if the parameter is present.

Additionally, with the Boolean variables, sending None through the gql calls overwrites whatever was in the transaction.  To avoid this, we have to ensure that it is not None.  Finally, the parameter is explicitly cast to a Boolean to ensure the true/false determination is left up to this code instead of the Monarch backend.  None, False and "" will cast to False.  Any numerical value, non-empty-string (including "False"), and True will cast to True.

Same logic for notes; we don't want None erasing existing content.

The remaining variables will be included in the variables on each call.  Testing indicates that an empty string will not overwrite these categories.